includes/REST_API/Embed_Controller.php

```diff
@@ -0,0 +1,318 @@
+<?php
+/**
+ * Class Embed_Controller
+ *
+ * @package   Google\Web_Stories
+ * @copyright 2020 Google LLC
+ * @license   https://www.apache.org/licenses/LICENSE-2.0 Apache License 2.0
+ * @link      https://github.com/google/web-stories-wp
+ */
+
+/**
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Web_Stories\REST_API;
+
+use DOMDocument;
+use DOMElement;
+use DOMNodeList;
+use DOMXpath;
+use Google\Web_Stories\Story_Post_Type;
+use WP_Error;
+use WP_Http;
+use WP_Post;
+use WP_REST_Controller;
+use WP_REST_Request;
+use WP_REST_Response;
+use WP_REST_Server;
+
+/**
+ * API endpoint to facilitate embedding web stories.
+ *
+ * Class Embed_Controller
+ */
+class Embed_Controller extends WP_REST_Controller {
+	/**
+	 * Constructor.
+	 */
+	public function __construct() {
+		$this->namespace = 'web-stories/v1';
+		$this->rest_base = 'embed';
+	}
+
+	/**
+	 * Registers routes for links.
+	 *
+	 * @see register_rest_route()
+	 *
+	 * @return void
+	 */
+	public function register_routes() {
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base,
+			[
+				[
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => [ $this, 'get_proxy_item' ],
+					'permission_callback' => [ $this, 'get_proxy_item_permissions_check' ],
+					'args'                => [
+						'url' => [
+							'description' => __( 'The URL for which to fetch embed data.', 'web-stories' ),
+							'type'        => 'string',
+							'format'      => 'uri',
+							'required'    => true,
+						],
+					],
+				],
+			]
+		);
+	}
+
+	/**
+	 * Callback for the Web Stories embed endpoints.
+	 *
+	 * Returns information about the given story.
+	 *
+	 * @param WP_REST_Request $request Full data about the request.
+	 *
+	 * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
+	 */
+	public function get_proxy_item( $request ) {
+		$url = untrailingslashit( $request['url'] );
+
+		if ( empty( $url ) ) {
```

Comment:
Most of this logic seems repeat of logic found in the links controller. Can we use a trait or another class that is extended.

Reply:
For reference, same question here from Miina:

https://github.com/google/web-stories-wp/pull/1249#discussion_r420020108

I don't think at this point it's worth optimizing this too much.

In the future this block & endpoint will probably be split up into its own package/plugin anyway, so might be best to keep things decoupled.

Reply:
I think we can break this off to another ticket.